src/test/scripts/functions/codegenalg/Algorithm_PCA.R

```diff
@@ -0,0 +1,108 @@
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+#
+#-------------------------------------------------------------
+
+#
+# This script performs Principal Component Analysis (PCA) on the given input data.
+#
+
+args <- commandArgs(TRUE)
+library("Matrix")
+
+A = readMM(paste(args[1], "A.mtx", sep=""));
+K = ncol(A);
+#ofmt = "CSV";
+projectData = 0;
+model = "";
+center = 0;
+scale = 0;
+output =  "/";
+
+evec_dominant = matrix(0, 1, 1);
+
+if (model != "") {
+  # reuse existing model to project data
+} else if (model == "") {
+    model = output;
+}
+
+  N = nrow(A);
+  D = ncol(A);
+
+  # perform z-scoring (centering and scaling)
+  if (center == 1) {
+    cm = colMean(A);
+    A  = A - matrix(1, nrow(A), 1) %*% cm;
+  }
+  if (scale == 1) {
+    cvars = (colSums(A^2));
+    if (center == 1){
+      cm = colMeans(A);
+      cvars = (cvars - N*(cm^2))/(N-1);
+    }
+    Azscored = A / sqrt(cvars);
+    A = Azscored;
+  }
+
+  # co-variance matrix
+  mu = matrix(1, nrow(A), 1) %*% colSums(A)/N;
+  C = (t(A) %*% A)/(N-1) - (N/(N-1))*t(mu) %*% mu;
+
+  # list data structure to mutli assign function values
+  list <- structure(NA,class="result")
+  "[<-.result" <- function(x,...,value) {
+    args <- as.list(match.call())
+    args <- args[-c(1:2,length(args))]
+    length(value) <- length(args)
+    for(i in seq(along=args)) {
+      a <- args[[i]]
+      if(!missing(a)) eval.parent(substitute(a <- v,list(a=a,v=value[[i]])))
+    }
+    x
+  }
+
+  # compute eigen vectors and values
+  list[evalues, evectors] = eigen(C);
+
+  decreasing_Idx = order(as.vector(evectors), decreasing=TRUE);
+  diagmat = table(seq(1,D), decreasing_Idx[1:D]);
+  # sorts eigen values by decreasing order
+  evalues = diagmat %*% evalues;
+  # sorts eigen vectors column-wise in the order of decreasing eigen values
+  evectors = evectors %*% diagmat;
+
+  # select K dominant eigen vectors
+  nvec = ncol(evectors);
+
+  eval_dominant = evalues[1:K, 1];
+  evec_dominant = evectors[,1:K];
+  Eps = 10^-8;
+  # the square root of eigen values
+  eval_stdev_dominant = sqrt(eval_dominant); # NaNs produced here!
+
+  write(as(eval_stdev_dominant, "CsparseMatrix"), paste(model+"/dominant.eigen.standard.deviations"), sep=",");
+  write(as(eval_dominant, "CsparseMatrix"), paste(model+"/dominant.eigen.values"), sep=",");
+  write(as(evec_dominant, "CsparseMatrix"), paste("/dominant.eigen.vectors"), sep="");
+}
```

Comment:
Hi @mboehm7 ,Can you try running this script, I am facing file open `permission denied` problem on my computer.

Reply:
Yeah, it looks like you're missing the directory name in the last write and paste in R works with comma not plus.

Reply:
Also change the sep argument from "," to "".